File: src/main/java/com/sleepwalker/sleeplib/client/drawable/ModalSprite.java
package com.sleepwalker.sleeplib.client.drawable;

import com.mojang.blaze3d.matrix.MatrixStack;
import com.mojang.blaze3d.systems.RenderSystem;
import com.sleepwalker.sleeplib.gg.essential.universal.UMatrixStack;
import net.minecraft.client.renderer.BufferBuilder;
import net.minecraft.client.renderer.Tessellator;
import net.minecraft.client.renderer.WorldVertexBufferUploader;
import net.minecraft.client.renderer.vertex.DefaultVertexFormats;
import net.minecraft.util.math.vector.Matrix4f;
import org.lwjgl.opengl.GL11;

import javax.annotation.Nonnull;
import java.awt.*;

public class ModalSprite implements Drawable {

    @Nonnull protected final TextureSource texture;
    protected final int uOffset;
    protected final int vOffset;
    protected final int uWidth;
    protected final int vHeight;

    public ModalSprite(@Nonnull TextureSource texture, int uOffset, int vOffset, int uWidth, int vHeight) {
        this.texture = texture;
        this.uOffset = uOffset;
        this.vOffset = vOffset;
        this.uWidth = uWidth;
        this.vHeight = vHeight;
    }

    @Override
    public void drawImage(@Nonnull UMatrixStack ms, double x, double y, double width, double height, @Nonnull Color color) {
        texture.bind();
        blit(ms.toMC(), color, (float) x, (float) y, 0, uOffset, vOffset, uWidth, vHeight);
    }

    @Override
    public int getWidth() {
        return uWidth;
    }

    @Override
    public int getHeight() {
        return vHeight;
    }

    protected void blit(@Nonnull MatrixStack ms, @Nonnull Color color, float pX, float pY, float zLevel, float pUOffset, float pVOffset, int pUWidth, int pVHeight) {
        innerBlit(ms, color, pX, pX + pUWidth, pY, pY + pVHeight, zLevel, pUWidth, pVHeight, pUOffset, pVOffset, texture.getWidth(), texture.getHeight());
    }

    public static void blit(@Nonnull MatrixStack ms, @Nonnull Color color, float pX, float pY, float zLevel, float pUOffset, float pVOffset, int pUWidth, int pVHeight, int pTextureWidth, int pTextureHeight) {
        innerBlit(ms, color, pX, pX + pUWidth, pY, pY + pVHeight, zLevel, pUWidth, pVHeight, pUOffset, pVOffset, pTextureWidth, pTextureHeight);
    }

    public static void blit(MatrixStack ms, @Nonnull Color color, float x, float y, float width, float height, float uOffset, float vOffset, int uWidth, int vHeight, int textureWidth, int textureHeight) {
        innerBlit(ms, color, x, x + width, y, y + height, 0, uWidth, vHeight, uOffset, vOffset, textureWidth, textureHeight);
    }

    public static void innerBlit(@Nonnull MatrixStack ms, @Nonnull Color color, float x1, float x2, float y1, float y2, float z, int pUWidth, int pVHeight, float pUOffset, float pVOffset, int pTextureWidth, int pTextureHeight) {
        innerBlit(ms.getLast().getMatrix(), color, x1, x2, y1, y2, z, (pUOffset + 0.0F) / (float)pTextureWidth, (pUOffset + (float)pUWidth) / (float)pTextureWidth, (pVOffset + 0.0F) / (float)pTextureHeight, (pVOffset + (float)pVHeight) / (float)pTextureHeight);
    }

    @SuppressWarnings("deprecation")
    public static void innerBlit(@Nonnull Matrix4f mat, @Nonnull Color color, float x1, float x2, float y1, float y2, float z, float minU, float maxU, float minV, float maxV) {
        float r = (float) color.getRed() / 255f;
        float g = (float) color.getGreen() / 255f;
        float b = (float) color.getBlue() / 255f;
        float a = (float) color.getAlpha() / 255f;

        BufferBuilder buffer = Tessellator.getInstance().getBuffer();
        buffer.begin(GL11.GL_QUADS, DefaultVertexFormats.POSITION_TEX_COLOR);
        buffer.pos(mat, x1, y2, z).tex(minU, maxV).color(r, g, b, a).endVertex();
        buffer.pos(mat, x2, y2, z).tex(maxU, maxV).color(r, g, b, a).endVertex();
        buffer.pos(mat, x2, y1, z).tex(maxU, minV).color(r, g, b, a).endVertex();
        buffer.pos(mat, x1, y1, z).tex(minU, minV).color(r, g, b, a).endVertex();
        buffer.finishDrawing();
        RenderSystem.enableAlphaTest();
        WorldVertexBufferUploader.draw(buffer);
    }
}

File: src/main/kotlin/com/sleepwalker/sleeplib/elementa/drawable/ModalSprite.kt
package com.sleepwalker.sleeplib.elementa.drawable

import com.sleepwalker.sleeplib.gg.essential.universal.UGraphics
import com.sleepwalker.sleeplib.gg.essential.universal.UMatrixStack
import org.lwjgl.opengl.GL11
import java.awt.Color

open class ModalSprite(
    val texture: TextureSource,
    val uOffset: Int,
    val vOffset: Int,
    val uWidth: Int,
    val vHeight: Int
) : Drawable {

    constructor(texture: TextureSource, uWidth: Int, vHeight: Int) : this(texture, 0, 0, uWidth, vHeight)
    constructor(texture: TextureSource) : this(texture, texture.width, texture.height)

    override val width: Int
        get() = uWidth
    override val height: Int
        get() = vHeight

    override fun drawImage(matrixStack: UMatrixStack, x: Double, y: Double, width: Double, height: Double, color: Color) {
        texture.bind()
        blit(matrixStack, color, x, y, 0.0, uOffset.toDouble(), vOffset.toDouble(), uWidth, vHeight)
    }

    protected fun blit(ms: UMatrixStack, color: Color, pX: Double, pY: Double, zLevel: Double, pUOffset: Double, pVOffset: Double, pUWidth: Int, pVHeight: Int) {
        innerBlit(ms, color, pX, pX + pUWidth, pY, pY + pVHeight, zLevel, pUWidth, pVHeight, pUOffset, pVOffset, texture.width, texture.height)
    }

    companion object {

        fun blit(ms: UMatrixStack, color: Color, pX: Double, pY: Double, zLevel: Double, pUOffset: Double, pVOffset: Double, pUWidth: Int, pVHeight: Int, pTextureWidth: Int, pTextureHeight: Int) {
            innerBlit(ms, color, pX, pX + pUWidth, pY, pY + pVHeight, zLevel, pUWidth, pVHeight, pUOffset, pVOffset, pTextureWidth, pTextureHeight)
        }

        fun blit(ms: UMatrixStack, color: Color, x: Double, y: Double, width: Double, height: Double, uOffset: Double, vOffset: Double, uWidth: Int, vHeight: Int, textureWidth: Int, textureHeight: Int) {
            innerBlit(ms, color, x, x + width, y, y + height, 0.0, uWidth, vHeight, uOffset, vOffset, textureWidth, textureHeight)
        }

        fun innerBlit(ms: UMatrixStack, color: Color, x1: Double, x2: Double, y1: Double, y2: Double, z: Double, pUWidth: Int, pVHeight: Int, pUOffset: Double, pVOffset: Double, pTextureWidth: Int, pTextureHeight: Int) {
            innerBlit(ms, color, x1, x2, y1, y2, z,
                (pUOffset + 0.0f) / pTextureWidth.toDouble(),
                (pUOffset + pUWidth.toDouble()) / pTextureWidth.toDouble(),
                (pVOffset + 0.0f) / pTextureHeight.toDouble(),
                (pVOffset + pVHeight.toDouble()) / pTextureHeight.toDouble()
            )
        }

        fun innerBlit(ms: UMatrixStack, color: Color, x1: Double, x2: Double, y1: Double, y2: Double, z: Double, minU: Double, maxU: Double, minV: Double, maxV: Double) {
            val r = color.red.toFloat() / 255f
            val g = color.green.toFloat() / 255f
            val b = color.blue.toFloat() / 255f
            val a = color.alpha.toFloat() / 255f

            val worldRenderer = UGraphics.getFromTessellator()
            worldRenderer.beginWithDefaultShader(UGraphics.DrawMode.QUADS, UGraphics.CommonVertexFormats.POSITION_TEXTURE_COLOR)
            worldRenderer.pos(ms, x1, y2, z).tex(minU, maxV).color(r, g, b, a).endVertex()
            worldRenderer.pos(ms, x2, y2, z).tex(maxU, maxV).color(r, g, b, a).endVertex()
            worldRenderer.pos(ms, x2, y1, z).tex(maxU, minV).color(r, g, b, a).endVertex()
            worldRenderer.pos(ms, x1, y1, z).tex(minU, minV).color(r, g, b, a).endVertex()

            UGraphics.enableBlend()
            UGraphics.enableAlpha()
            UGraphics.enableDepth()
            UGraphics.tryBlendFuncSeparate(GL11.GL_SRC_ALPHA, GL11.GL_ONE_MINUS_SRC_ALPHA, GL11.GL_ONE, GL11.GL_ZERO)
            worldRenderer.drawDirect()
        }
    }
}